Fix deprecations reported on Gradle 8.5

britter opened the pull request:
Summary of the changes:

- Use GradleVersion instead of VersionNumber (same as on master branch)
- Use `getMainClass()` on `JavaApplication` extension
- Call `getSourceSets()` via reflection on `JavaPluginExtension` when running on a Gradle version that supports it.
- Add a test verifying configuring the plugin on Gradle 8.5 does not yield deprecation warning.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/ratpack/ratpack/1704)
<!-- Reviewable:end -->

britter committed:
Use GradleVersion class instead of VersionNumber

VersionNumber is deprecated and will be removed with Gradle 9.x.
This change was already applied on master in cb4eb1b420ed8995f07d81ff417a2dd8bc62d4ea.

britter committed:
Fix deprecations in ratpack-gradle reported on Gradle 8.5

britter commented:
@ldaley looking at the logs this could be a temporary issue with reactor JavaDocs: https://app.circleci.com/pipelines/github/ratpack/ratpack/693/workflows/536d7a0f-8a1c-4e63-aabb-c97c41143958/jobs/1599?invite=true#step-103-125153_85 Can you try re-running?

ldaley commented:
Rerunning.

ldaley commented:
@britter looks like there are legit test failures. 

My kingdom for a build scan.

britter committed:
More nuanced handling of mainClass properties

The methods JavaExec.getMain and JavaApplication.getMainClass have been
removed in Gradle 8.0. Their replacements had been introduced in
Gradle 6.4. Due to the fact that the shadow plugin kept using the old
versions until shadow plugin release 6.0, ratpack-gradle keeps using
them as well, if running below Gradle 8.0 in order to not break
integration with the shadow plugin.

britter commented:
@ldaley this should work now.

ldaley commented:
Thank you @britter
